pulp_ansible/app/sigstore/issuers/keycloak.py

```diff
@@ -0,0 +1,135 @@
+"""
+Representation of Keycloak as an OIDC issuer.
+"""
+
+from sigstore.oidc import _OpenIDConfiguration
+from sigstore.oidc import IdentityError
+
+import logging
+import os
+import requests
+import time
+import webbrowser
+
+from pulp_ansible.app.sigstore.exceptions import KeycloakException
+from pulp_ansible.app.sigstore.issuers.base import BaseIssuer
+
+
+log = logging.getLogger(__name__)
+
+
+class _KeycloakOpenIDConfiguration(_OpenIDConfiguration):
+    """
+    Extends sigstore-python's _OpenIDConfiguration
+    with checks relative to Keycloak configuration.
+    """
+
+
+class Keycloak(BaseIssuer):
+    """
+    Custom representation of Keycloak as an OIDC Issuer.
+    Extends the functionalities provided by
+    https://github.com/sigstore/sigstore-python/blob/v1.1.0/sigstore/oidc.py#L55
+    """
+
+    def __init__(self, keycloak_base_url):
+        """Create a new Keycloak issuer with custom endpoints."""
+        oidc_config_url = os.path.join(keycloak_base_url, ".well-known/openid-configuration")
+
+        resp: requests.Response = requests.get(oidc_config_url)
+        try:
+            resp.raise_for_status()
+        except requests.HTTPError as http_error:
+            log.error(http_error.response)
+            raise KeycloakException from http_error
+
+        try:
+            self.oidc_config = _KeycloakOpenIDConfiguration.parse_obj(resp.json())
+        except ValueError as exc:
+            raise KeycloakException(f"Keycloak returned invalid configuration: {exc}")
+
+    def identity_token(self, client_id, client_secret, enable_interactive):
+        """Get an identity token from Keycloak token endpoint."""
+        # interactive mode is taken from the original sigstore python
+        # Issuer.identity_token() implementation at
+        # https://github.com/sigstore/sigstore-python/blob/v1.1.0/sigstore/oidc.py#L100
+        # This method works only on browser interaction and enables out-of-bond
+        # only if the former method fails.
+        if enable_interactive:
+            from sigstore._internal.oidc.oauth import _OAuthFlow
+
+            code: str
+            with _OAuthFlow(client_id, client_secret, self) as server:
+                # Launch web browser
+                if webbrowser.open(server.base_uri):
```

Comment:
This code simply won't work for the signing task as 99% of the time the machine the task is running on is not accessible by the user, so the signing task will simply stall until it fails waiting for the user input. 

I think I have a good idea for how to get around this problem, but it's pretty in-depth and will take me a while to write down. I would also like to get feedback from my other team-members, so let's discuss it at our next sync.

Reply:
Thanks for your input during the sync, I removed the interactive signing part in https://github.com/pulp/pulp_ansible/pull/1428/commits/ca80190904420239b6fbbc0992ba3df935da91f8